load-test/cantabular/autocannon-go-2/INSTRUCTIONS.md

```diff
@@ -0,0 +1,155 @@
+```
+
+from the first one above, in what it outputs we pick the large and 'real' databases and get a list of each's codebook with:
+
+```shell
+curl 'http://cantabular-server-publishing.internal.staging:14000/v10/codebook/People-Households?cats=false' | jq .
+
+curl 'http://cantabular-server-publishing.internal.staging:14000/v10/codebook/Usual-Residents?cats=false' | jq .
+
+curl 'http://cantabular-server-publishing.internal.staging:14000/v10/codebook/Household-Ref-Persons?cats=false' | jq .
+```
+
+The output from the above is then built into the array `uriList` in the function `runClients()` in the structure you see in the code.
+
+## Details of load test
+
+How much load is applied is tunable, but requires real observations to see effect of changing parameters within the Makefile and the file autocannon.go
+
+Load is applied to the Cantabular servers in Staging Publishing by having an instance of the autocannon.go built executable (for AWS machines) run from one or more of the publishing boxes in Staging (so as to be able to access the Cantabular Servers in Staging Publishing network).
+
+The load test is observed by configuring the Cantabular Publishing Server ASG in staging to have a min of three machines and a max of six (instance type r5.large, at time of initial testing), and then logging into each of the 3 servers in seperate terminals and running `htop` in each.
+
+## Running load test
+
+Ensure you have selected the staging environment for AWS.
+
+The load test is then run by simply doing: `make`
+
+Shortly after the `make` has finished, the load tests will start running on the publishing boxes and apply load to the Cantabular publishing cluster in staging.
+
+The 'make' will build the autocannon.go app to run on linux, load it into  1 or more publishing boxes (as determined by `host_numbers` list in the Makefile) and cause each of these apps to start running some minutes later as determined by lines like this in the Makefile:
+
+```makefile
+TZ=UTC; start_time=$(shell TZ=UTC date -v +3M '+%H:%M'); \
+```
+
+(where the `+3M` is three minutes, which is OK for 4 instances of autocannon.go being deployed ... This `+3M` needs changing to `+4M` for six instances and needs to be `+5M` for 8 instances)
+
+The time delay for execution gives the makefile enough time to sequentially issue the commands to each publishing box, and then the apps all run at the same time to apply load.
+
+Observe the effect of the load test with `htop` on the Cantabular server boxes. This will allow you to see when the load testin is happening and when it has finished.
+
+## Adjusting amount of load
+
+The load applied can be increased or decreased in a number of ways, where the aim is to first run the load test with the `make` command and observe the current load via `htop` for 3 servers over the duration of the test.
+
+Next we adjust the value for `pipelining` in this line of code:
```

Comment:
What does pipelining mean?

Reply:
Search for that word in the code and see how its used.